crates/chain/src/tx_graph.rs

```diff
@@ -598,6 +598,31 @@ impl<A: Clone + Ord> TxGraph<A> {
 }
 
 impl<A: Anchor> TxGraph<A> {
+    /// Find missing block heights of `chain`.
+    ///
+    /// This works by scanning through anchors, and seeing whether the anchor block of the anchor
+    /// exists in the [`LocalChain`].
+    pub fn missing_blocks<'a>(&'a self, chain: &'a LocalChain) -> impl Iterator<Item = u32> + 'a {
+        self.anchors
+            .iter()
+            .map(|(a, _)| a.anchor_block())
+            .filter({
+                let mut last_block = Option::<BlockId>::None;
+                move |block| {
+                    if last_block.as_ref() == Some(block) {
+                        false
+                    } else {
+                        last_block = Some(*block);
+                        true
+                    }
+                }
+            })
+            .filter_map(|block| match chain.heights().get(&block.height) {
+                Some(chain_hash) if *chain_hash == block.hash => None,
+                _ => Some(block.height),
+            })
+    }
```

Comment:
I recall in the previous PR that this function was implemented incorrectly and I requested tests be written. I re-requesting this. I still believe this function is wrong. it needs tests for:

1. Two transactions with anchors the same anchor which is missing from the chain: the missing block appears once in the iterator.
2. Two transactions with *different* anchors at the same height: the missing block appears once in the chain.
3. One transaction with two anchors at the same height. The chain doesn't include this height. The iterator should include one item.
4. One transaction with two anchors at the same height. The chain includes this height but doesn't match either anchor. The iterator should be empty.
5. One transaction with two anchors at two different heights. One of the anchor height is included in the chain with the same hash as the anchor. This should return nothing.
6. One transaction with two anchors at two different heights A and B. height  A is included in the chain with a different hash to as the anchor. The iterator should yield only B.
7. One transaction with two anchors at two different heights. Neither height is in the chain. This should yield both heights in the iterator.

The essential principle is that it should return any height that could help it determine whether one of the transactions in the graph is in the chain or not. If it can determine that already then it shouldn't return anything for that tx.

Reply:
> One transaction with two anchors at two different heights A and B. height A is included in the chain with a different hash to as the anchor. The iterator should yield only B.

Isn't it possible that the transaction is confirmed at height A anyway? My intuition tells me we should still return both heights.

Reply:
@LLFourn I think I did it, except for the test case I don't agree with. 4a704fe700cc608a4760173be1d4a9a506a2b73a